File: src/api/app/security.py
"""Security helpers for password hashing and JWT handling."""

import base64
import hashlib
import hmac
import json
import secrets
import time
import uuid
from datetime import UTC, datetime, timedelta
from typing import Any

from app.config import settings


class TokenDecodeError(Exception):
    """Raised when a token cannot be decoded or validated."""


def _b64url_encode(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).decode("utf-8").rstrip("=")


def _b64url_decode(raw: str) -> bytes:
    padding = "=" * ((4 - len(raw) % 4) % 4)
    return base64.urlsafe_b64decode(raw + padding)


def _json_compact(payload: dict[str, Any]) -> bytes:
    return json.dumps(payload, separators=(",", ":"), sort_keys=True).encode("utf-8")


def _encode_jwt(payload: dict[str, Any]) -> str:
    header = {"alg": "HS256", "typ": "JWT"}
    header_part = _b64url_encode(_json_compact(header))
    payload_part = _b64url_encode(_json_compact(payload))
    signing_input = f"{header_part}.{payload_part}".encode()
    signature = hmac.new(
        settings.API_JWT_SECRET.encode("utf-8"),
        signing_input,
        hashlib.sha256,
    ).digest()
    signature_part = _b64url_encode(signature)
    return f"{header_part}.{payload_part}.{signature_part}"


def _decode_jwt(token: str) -> dict[str, Any]:
    parts = token.split(".")
    if len(parts) != 3:
        raise TokenDecodeError("Malformed token.")

    header_part, payload_part, signature_part = parts
    signing_input = f"{header_part}.{payload_part}".encode()
    expected_signature = hmac.new(
        settings.API_JWT_SECRET.encode("utf-8"),
        signing_input,
        hashlib.sha256,
    ).digest()

    provided_signature = _b64url_decode(signature_part)
    if not hmac.compare_digest(expected_signature, provided_signature):
        raise TokenDecodeError("Invalid token signature.")

    payload_bytes = _b64url_decode(payload_part)
    payload = json.loads(payload_bytes.decode("utf-8"))
    if not isinstance(payload, dict):
        raise TokenDecodeError("Invalid payload.")

    exp = payload.get("exp")
    if not isinstance(exp, int):
        raise TokenDecodeError("Missing expiration.")
    if exp <= int(time.time()):
        raise TokenDecodeError("Token expired.")

    return payload


def hash_password(password: str) -> str:
    """Hash password with scrypt."""

    if len(password) < 8:
        raise ValueError("Password must be at least 8 characters.")

    n, r, p = 2**14, 8, 1
    salt = secrets.token_bytes(16)
    derived = hashlib.scrypt(
        password.encode("utf-8"),
        salt=salt,
        n=n,
        r=r,
        p=p,
        dklen=64,
    )
    return (
        "scrypt"
        f"${n}"
        f"${r}"
        f"${p}"
        f"${_b64url_encode(salt)}"
        f"${_b64url_encode(derived)}"
    )


def verify_password(password: str, encoded_hash: str | None) -> bool:
    """Verify password against stored scrypt hash."""

    if not encoded_hash:
        return False
    try:
        algorithm, n_raw, r_raw, p_raw, salt_raw, digest_raw = encoded_hash.split("$")
        if algorithm != "scrypt":
            return False
        n, r, p = int(n_raw), int(r_raw), int(p_raw)
        salt = _b64url_decode(salt_raw)
        expected = _b64url_decode(digest_raw)
    except (ValueError, TypeError):
        return False

    actual = hashlib.scrypt(
        password.encode("utf-8"),
        salt=salt,
        n=n,
        r=r,
        p=p,
        dklen=len(expected),
    )
    return hmac.compare_digest(actual, expected)


def hash_refresh_token(token: str) -> str:
    """Hash refresh token for storage."""

    return hashlib.sha256(token.encode("utf-8")).hexdigest()


def issue_access_token(user_id: uuid.UUID) -> tuple[str, int]:
    """Create a signed access token."""

    now = datetime.now(UTC)
    ttl_seconds = settings.API_ACCESS_TOKEN_EXPIRE_MINUTES * 60
    exp = int((now + timedelta(seconds=ttl_seconds)).timestamp())
    payload = {
        "iss": settings.API_JWT_ISSUER,
        "sub": str(user_id),
        "typ": "access",
        "iat": int(now.timestamp()),
        "exp": exp,
    }
    return _encode_jwt(payload), ttl_seconds


def issue_refresh_token(user_id: uuid.UUID, session_id: uuid.UUID) -> tuple[str, datetime]:
    """Create a signed refresh token."""

    now = datetime.now(UTC)
    expires_at = now + timedelta(days=settings.API_REFRESH_TOKEN_EXPIRE_DAYS)
    payload = {
        "iss": settings.API_JWT_ISSUER,
        "sub": str(user_id),
        "sid": str(session_id),
        "typ": "refresh",
        "iat": int(now.timestamp()),
        "exp": int(expires_at.timestamp()),
    }
    return _encode_jwt(payload), expires_at


def decode_access_token(token: str) -> dict[str, Any]:
    """Decode and validate access token."""

    payload = _decode_jwt(token)
    if payload.get("typ") != "access":
        raise TokenDecodeError("Unexpected token type.")
    return payload


def decode_refresh_token(token: str) -> dict[str, Any]:
    """Decode and validate refresh token."""

    payload = _decode_jwt(token)
    if payload.get("typ") != "refresh":
        raise TokenDecodeError("Unexpected token type.")
    return payload

File: apps/api/src/lib/security.ts
import { createHmac, randomBytes, scryptSync, timingSafeEqual } from "node:crypto";

import type { AppConfig } from "./config.js";

export class TokenDecodeError extends Error {}

const b64UrlEncode = (value: Buffer): string => value.toString("base64url").replace(/=/g, "");

const b64UrlDecode = (value: string): Buffer => Buffer.from(value, "base64url");

const jsonCompact = (payload: Record<string, unknown>): Buffer => {
  const sortedEntries = Object.entries(payload).sort(([left], [right]) =>
    left.localeCompare(right)
  );
  return Buffer.from(JSON.stringify(Object.fromEntries(sortedEntries)));
};

const hashWithPurpose = (token: string, purpose: string, secret: string): string =>
  createHmac("sha256", secret).update(`${purpose}:${token}`).digest("hex");

const encodeJwt = <T extends object>(payload: T, secret: string): string => {
  const headerPart = b64UrlEncode(jsonCompact({ alg: "HS256", typ: "JWT" }));
  const payloadPart = b64UrlEncode(jsonCompact(payload as Record<string, unknown>));
  const signingInput = `${headerPart}.${payloadPart}`;
  const signature = createHmac("sha256", secret).update(signingInput).digest();
  return `${signingInput}.${b64UrlEncode(signature)}`;
};

const decodeJwt = (token: string, secret: string): Record<string, unknown> => {
  const parts = token.split(".");
  if (parts.length !== 3) {
    throw new TokenDecodeError("Malformed token.");
  }

  const [headerPart, payloadPart, signaturePart] = parts;
  const signingInput = `${headerPart}.${payloadPart}`;
  const expectedSignature = createHmac("sha256", secret).update(signingInput).digest();
  const providedSignature = b64UrlDecode(signaturePart);

  if (
    expectedSignature.length !== providedSignature.length ||
    !timingSafeEqual(expectedSignature, providedSignature)
  ) {
    throw new TokenDecodeError("Invalid token signature.");
  }

  const payload = JSON.parse(b64UrlDecode(payloadPart).toString("utf8"));
  if (typeof payload !== "object" || payload === null || Array.isArray(payload)) {
    throw new TokenDecodeError("Invalid payload.");
  }

  const exp = payload.exp;
  if (typeof exp !== "number") {
    throw new TokenDecodeError("Missing expiration.");
  }

  if (exp <= Math.floor(Date.now() / 1000)) {
    throw new TokenDecodeError("Token expired.");
  }

  return payload as Record<string, unknown>;
};

export const hashPassword = (password: string): string => {
  if (password.length < 8) {
    throw new Error("Password must be at least 8 characters.");
  }

  const n = 2 ** 14;
  const r = 8;
  const p = 1;
  const salt = randomBytes(16);
  const derived = scryptSync(password, salt, 64, {
    N: n,
    r,
    p,
  });

  return `scrypt$${n}$${r}$${p}$${b64UrlEncode(salt)}$${b64UrlEncode(derived)}`;
};

export const verifyPassword = (password: string, encodedHash: string | null): boolean => {
  if (!encodedHash) {
    return false;
  }

  try {
    const [algorithm, nRaw, rRaw, pRaw, saltRaw, digestRaw] = encodedHash.split("$");
    if (algorithm !== "scrypt") {
      return false;
    }

    const expected = b64UrlDecode(digestRaw);
    const actual = scryptSync(password, b64UrlDecode(saltRaw), expected.length, {
      N: Number.parseInt(nRaw, 10),
      r: Number.parseInt(rRaw, 10),
      p: Number.parseInt(pRaw, 10),
    });

    return timingSafeEqual(actual, expected);
  } catch {
    return false;
  }
};

export const hashRefreshToken = (
  token: string,
  config: Pick<AppConfig, "API_TOKEN_HASH_SECRET">
): string => hashWithPurpose(token, "refresh", config.API_TOKEN_HASH_SECRET);

export const hashToken = (
  token: string,
  config: Pick<AppConfig, "API_TOKEN_HASH_SECRET">
): string => hashWithPurpose(token, "generic", config.API_TOKEN_HASH_SECRET);

export interface AccessClaims {
  iss: string;
  sub: string;
  typ: "access";
  iat: number;
  exp: number;
}

export interface RefreshClaims {
  iss: string;
  sub: string;
  sid: string;
  typ: "refresh";
  iat: number;
  exp: number;
}

export const issueAccessToken = (
  userId: string,
  config: AppConfig
): { token: string; expiresIn: number } => {
  const now = Math.floor(Date.now() / 1000);
  const expiresIn = config.API_ACCESS_TOKEN_EXPIRE_MINUTES * 60;
  const payload: AccessClaims = {
    iss: config.API_JWT_ISSUER,
    sub: userId,
    typ: "access",
    iat: now,
    exp: now + expiresIn,
  };

  return {
    token: encodeJwt(payload, config.API_JWT_SECRET),
    expiresIn,
  };
};

export const issueRefreshToken = (
  userId: string,
  sessionId: string,
  config: AppConfig
): { token: string; expiresAt: Date } => {
  const now = Math.floor(Date.now() / 1000);
  const expiresAt = new Date(Date.now() + config.API_REFRESH_TOKEN_EXPIRE_DAYS * 86400_000);
  const payload: RefreshClaims = {
    iss: config.API_JWT_ISSUER,
    sub: userId,
    sid: sessionId,
    typ: "refresh",
    iat: now,
    exp: Math.floor(expiresAt.getTime() / 1000),
  };

  return {
    token: encodeJwt(payload, config.API_JWT_SECRET),
    expiresAt,
  };
};

export const decodeAccessToken = (token: string, config: AppConfig): AccessClaims => {
  const payload = decodeJwt(token, config.API_JWT_SECRET);
  if (payload.typ !== "access") {
    throw new TokenDecodeError("Unexpected token type.");
  }
  return payload as unknown as AccessClaims;
};

export const decodeRefreshToken = (token: string, config: AppConfig): RefreshClaims => {
  const payload = decodeJwt(token, config.API_JWT_SECRET);
  if (payload.typ !== "refresh") {
    throw new TokenDecodeError("Unexpected token type.");
  }
  return payload as unknown as RefreshClaims;
};

const LISTING_VERIFY_PURPOSE = "listing-verify";
const LISTING_OWNERSHIP_TYP = "listing_ownership" as const;
const LISTING_OWNERSHIP_TTL_SECONDS = 15 * 60;

export interface ListingVerificationTarget {
  userId: string;
  host: string;
  port: number;
}

/**
 * Derives the deterministic ownership-verification token a server owner pastes
 * into their MOTD or a DNS TXT record. The same (user, host, port) always maps
 * to the same token, so it can be shown and re-derived without persistence.
 */
export const deriveListingVerificationToken = (
  target: ListingVerificationTarget,
  config: Pick<AppConfig, "API_TOKEN_HASH_SECRET">
): string => {
  const digest = createHmac("sha256", config.API_TOKEN_HASH_SECRET)
    .update(`${LISTING_VERIFY_PURPOSE}:${target.userId}:${target.host}:${target.port}`)
    .digest("hex");
  return `fs-verify-${digest.slice(0, 12).toUpperCase()}`;
};

export interface ListingOwnershipClaims {
  iss: string;
  sub: string;
  typ: typeof LISTING_OWNERSHIP_TYP;
  host: string;
  port: number;
  iat: number;
  exp: number;
}

/**
 * Issues a short-lived signed proof that a user has verified ownership of a
 * specific host:port. The create-listing endpoint requires this proof so that
 * verification cannot be skipped or replayed against a different listing.
 */
export const issueListingOwnershipProof = (
  target: ListingVerificationTarget,
  config: AppConfig
): { token: string; expiresIn: number } => {
  const now = Math.floor(Date.now() / 1000);
  const payload: ListingOwnershipClaims = {
    iss: config.API_JWT_ISSUER,
    sub: target.userId,
    typ: LISTING_OWNERSHIP_TYP,
    host: target.host,
    port: target.port,
    iat: now,
    exp: now + LISTING_OWNERSHIP_TTL_SECONDS,
  };

  return {
    token: encodeJwt(payload, config.API_JWT_SECRET),
    expiresIn: LISTING_OWNERSHIP_TTL_SECONDS,
  };
};

/** Validates and decodes a listing ownership proof, enforcing signature, type, and expiry. */
export const decodeListingOwnershipProof = (
  token: string,
  config: AppConfig
): ListingOwnershipClaims => {
  const payload = decodeJwt(token, config.API_JWT_SECRET);
  if (payload.typ !== LISTING_OWNERSHIP_TYP) {
    throw new TokenDecodeError("Unexpected token type.");
  }
  return payload as unknown as ListingOwnershipClaims;
};
